faiss-search/app.py

```diff
@@ -0,0 +1,102 @@
+    return os.path.join(workspace_path, dn)
+
+
+def read_data(db_file_path: str) -> Generator[bytes, Any, Any]:
+    vectors = fvecs_read(db_file_path)
+    num_vectors = vectors.shape[0]
+    for i in range(1, num_vectors + 1):
+        start_batch = (i - 1)
+        end_batch = i if i < num_vectors else num_vectors
+        yield vectors[start_batch: end_batch].tobytes()
+
+
+def save_topk(resp, output_file, top_k):
+    results = []
+    with open(output_file, 'w') as fw:
+        query_id = 0
+        for d in resp.search.docs:
+            result = []
+            fw.write('-' * 20)
+            fw.write('\n')
+            fw.write('query id {}: \n {}'.format(query_id, np.frombuffer(d.blob.buffer, d.blob.dtype)))
+            fw.write('\n')
+            fw.write('matched vectors' + "*" * 10)
+            fw.write('\n')
+            for idx, kk in enumerate(d.topk_results):
+                result.append(kk.match_doc.doc_id)
+                score = kk.score.value
+                if score < 0.0:
+                    continue
+                m_fn = np.frombuffer(kk.match_doc.blob.buffer, kk.match_doc.blob.dtype)
+                fw.write('\n')
+                fw.write('Idx: {:>2d}:(DocId {}, Ranking score: {:f}): \n{}'.
+                         format(idx, kk.match_doc.doc_id, score, m_fn))
+                fw.write('\n')
+            fw.write('\n')
+            results.append(result)
+            query_id += 1
+        fw.write(f'recall@{top_k}: {recall_at_k(np.array(results), top_k)}')
+
+    print(open(output_file, 'r').read())
+
+
+def recall_at_k(results, k):
+    """
+    Computes how many times the true nearest neighbour is returned as one of the k closest vectors from a query.
+
+    Taken from https://gist.github.com/mdouze/046c1960bc82801e6b40ed8ee677d33e
+    """
+    groundtruth_path = os.path.join(os.environ['TMP_DATA_DIR'], 'siftsmall_groundtruth.ivecs')
+    groundtruth = ivecs_read(groundtruth_path)
+    eval = (results[:, :k] == groundtruth[:, :1]).sum() / float(results.shape[0])
+    return eval
+
+
+@click.command()
+@click.option('--task', '-t')
+@click.option('--batch_size', '-n', default=50)
+@click.option('--top_k', '-k', default=5)
+def main(task, batch_size, top_k):
+    os.environ['TMP_WORKSPACE'] = get_random_ws(os.environ['TMP_DATA_DIR'])
+    if task == 'index':
+        data_path = os.path.join(os.environ['TMP_DATA_DIR'], 'siftsmall_base.fvecs')
+        flow = Flow().load_config('flow-index.yml')
+        with flow.build() as fl:
+            fl.index(read_data(data_path), batch_size=batch_size)
+    elif task == 'query':
+        data_path = os.path.join(os.environ['TMP_DATA_DIR'], 'siftsmall_query.fvecs')
+        flow = Flow().load_config('flow-query.yml')
+        with flow.build() as fl:
+            ppr = lambda x: save_topk(x, os.path.join(os.environ['TMP_DATA_DIR'], 'query_results.txt'), top_k)
+            fl.search(buffer=read_data(data_path), callback=ppr, top_k=top_k)
```

Comment:
`callback` will be deprecated
```suggestion
            fl.search_ndarray(read_data(data_path), output_fn=ppr, top_k=top_k)
```